### PR TITLE
PLNSRVCE-1164: add new CRD to capture proxy plugin configuration for the registration service

### PR DESCRIPTION
**Pull request by gabemontero**
## Description
In the context of stonesoup, and the desire to allow the HAC to make tekton results API calls through the host operator tier via the registration service to the member cluster where a given pipeline runs was executed, this PR is introducing a new CR for capturing the necessary config to looking up the tekton results endpoint (in this case an openshift route) on the member cluster.

## Checks
1. Did you run `make generate` target? **yes**

2. Did `make generate` change anything in other projects (host-operator, member-operator)? **yes**

3. In case of **new** CRD, did you the following? **yes/no**
    - update make/generate.mk in this repository **yes**
    - `resources/setup/roles/host.yaml` in the sandbox-sre repository **not yet, but @alexeykazakov has given me access to that cluster
    - `PROJECT` file: https://github.com/codeready-toolchain/host-operator/blob/master/PROJECT **yes**
    - `CSV` file: https://github.com/codeready-toolchain/host-operator/blob/master/config/manifests/bases/host-operator.clusterserviceversion.yaml **yes**

4. In case other projects are changed, please provides PR links.
    - host-operator: https://github.com/codeready-toolchain/host-operator/pull/#758
    - member-operator: https://github.com/codeready-toolchain/member-operator/pull/#

### CODE REVIEW

File: api/v1alpha1/proxyplugin_types.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+const (
+	OpenShiftRouteEndpointType = ProxyPluginEndpointType("OpenShiftRouteEndpointType")
+)
+
+type ProxyPluginEndpointType string
+
+type ProxyPluginEntry struct {
+	// PluginName identifies the specific plugin, and is the next segment after '/plugin' in the request
```
Comment (alexeykazakov) [inline api/v1alpha1/proxyplugin_types.go:12]: I think we should be using `.../plugins/<pluginName>/...` instead. Like `/workspaces/<workspaceName>/...` or `/namespaces/<namespaceName>/...`
```suggestion
	// PluginName identifies the specific plugin, and is the next segment after '/plugins/' in the request: '/plugins/<pluginName>/...'
```

File: api/v1alpha1/proxyplugin_types.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+const (
+	OpenShiftRouteEndpointType = ProxyPluginEndpointType("OpenShiftRouteEndpointType")
```
Comment (alexeykazakov) [inline api/v1alpha1/proxyplugin_types.go:6]: Maybe just
```suggestion
	OpenShiftRouteEndpointType = ProxyPluginEndpointType("OpenShiftRoute")
```
?

File: api/v1alpha1/proxyplugin_types.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+type ProxyPluginEndpointType string
+
+type ProxyPluginEntry struct {
+	// PluginName identifies the specific plugin, and is the next segment after '/plugin' in the request
+	// URL that arrives to the registration service
+	PluginName string `json:"pluginName"`
+
+	// EndpointType designates which type of k8s Object in the member cluster represents the HTTP host destination
+	// of the proxied reqeust.  If this field is not set, the default of OpenShiftRouteEndpointType is assumed.
+	EndpointType ProxyPluginEndpointType `json:"endpointType,omitempty"`
```
Comment (alexeykazakov) [inline api/v1alpha1/proxyplugin_types.go:18]: Maybe `TargetEndpointType/Namespace/Name`?

File: api/v1alpha1/proxyplugin_types.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// ProxyPluginEntries is an array of proxy plugins configured for the registration service
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	ProxyPluginEntries []ProxyPluginEntry `json:"proxyPluginEntries,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
Comment (alexeykazakov) [inline api/v1alpha1/proxyplugin_types.go:39]: Do you think it's better to support multiple plugins in a single `ProxypPlugin`?

Wouldn't it be more straightforward to keep one CR per plugin?

```
kind: ProxyPlugin
metadata:
  name: tekton-results
spec:
  targetEndpointType: OpenShiftRoute
  targetEndpointNamespace: pipelines-operator
  targetEndpointName: tekton-result
```

Also, what do you think about having something like:
```
kind: ProxyPlugin
metadata:
  name: tekton-results
spec:
  openShiftRoutTargetEndpoint:
    namespace: pipelines-operator
    routeName: tekton-result
```
Where `openShiftRoutTargetEndpoint` is optional.
Later, if we want to support more endpoints types we add more dedicated types like `ingressTargetEndpoint` or something else with specific sub-fiels. And only one type would be expected.

WDYT?

File: api/v1alpha1/proxyplugin_types.go (hunk 1 above)
Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:12]: yeah I tried to subset the url explanation and as a result this comment is unclear

I'm going to update the godoc comment to more directly quote what is in the feature refinement doc for tekton results https://docs.google.com/document/d/1_VanTqldsChZXhUEG7n4gMyKiOL_lGGTNNIQM6xe6Ag/edit#heading=h.nhnim9xopxlz

That we support these 2 forms of URLs:
```
https://<proxy-host>/plugins/tekton-results/v1alpha2/<namespace-name>/<result-api> 
or 
https://<proxy-host>/plugins/tekton-results/workspaces/<workspace-name>/v1alpha2/<namespace-name>
```

and that `PluginName` corresponds to `tekton-results` in those 2 examples.

Of course, if you happen to see this @alexeykazakov before I update the PR and have further comment, I'll try to factor that in before my next push to the PR.

thanks

File: api/v1alpha1/proxyplugin_types.go (hunk 2 above)
Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:6]: part of next push

File: api/v1alpha1/proxyplugin_types.go (hunk 3 above)
Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:18]: part of next push

File: api/v1alpha1/proxyplugin_types.go (hunk 4 above)
Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:39]: yeah I came at it with a minimize etcd impact, k8s object count, and have a global entity vs. a per plugin config object, as a lot of my recent OCP customer interactions when dealing with new CRDs, that has been a concern

but perhaps that was an incorrect assumption to bring into what you all are dealing with in sandbox, as the "customer" here

certainly in hindsight, as we discussed in the video conf, although we are designing for extensibility, tekton-results is in fact the only tangible use case currently on the books, so the realistic amount of these won't be large for the foreseeable future

so sure, I'll pivot toward this model instead @alexeykazakov assuming my etcd cost note does not sway you

and yes I'll pivot toward the 
```
  openShiftRoutTargetEndpoint:
    namespace: pipelines-operator
    routeName: tekton-result
```
form .... a la the different build strategies for openshift builds, if you are familiar

File: api/v1alpha1/proxyplugin_types.go (hunk 2 above)
Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:6]: actually we'll nix this in favor of 
```
kind: ProxyPlugin
metadata:
  name: tekton-results
spec:
  openShiftRoutTargetEndpoint:
    namespace: pipelines-operator
    routeName: tekton-result
```

**Comment by gabemontero** [comment]
ok PR rebased and updates pushed @alexeykazakov thanks

File: api/v1alpha1/proxyplugin_types.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+// ProxyPluginSpec defines the desired state of ProxyPlugin
+// +k8s:openapi-gen=true
+type ProxyPluginSpec struct {
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// OpenShiftRouteTargetEndpoint is an optional field that represents the look up information for an OpenShift Route
+	// as the endpoint for the registration service to proxy requests to that have the https://<proxy-host>/plugins/<ProxyPlugin.ObjectMeta.Name>
+	// in its incoming URL.  As we add more types besides OpenShift Routes, we will add more optional fields to this spec
+	// object
```
Comment (alexeykazakov) [inline api/v1alpha1/proxyplugin_types.go:20]: To be honest it's still a bit unclear to me what is the proper way of defining optional fields. But I personally always add `// +optional` to optional field. Not sure if it's really required if it's a pointer... ;) According to https://github.com/codeready-toolchain/host-operator/pull/758/files the CRD was generated correctly. But anyway, just wanted to mention it.
```suggestion
	// object
	// +optional
```

**Review by alexeykazakov — APPROVED** [review]
Looks good! Just one comment/question...

File: api/v1alpha1/proxyplugin_types.go (hunk 5 above)
Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:20]: I'll pull this commit down @alexeykazakov and re-run `make generate`

Comment (gabemontero) [reply api/v1alpha1/proxyplugin_types.go:20]: `make generate` did not produce any changes 

```
gmontero ~/go/src/github.com/codeready-toolchain/api  (add-proxy-config)$ git log -3
commit c4878af3a4fc0ab73bea4fae648eadc04b8a3c51 (HEAD -> add-proxy-config, origin/add-proxy-config)
Author: Gabe Montero <gmontero@redhat.com>
Date:   Fri Mar 3 16:52:18 2023 -0500

    Update api/v1alpha1/proxyplugin_types.go
    
    Co-authored-by: Alexey Kazakov <alkazako@redhat.com>

commit ba9ecad6b1ee86c6b1a18ec544f65aae3c3d547a
Author: gabemontero <gmontero@redhat.com>
Date:   Thu Mar 2 13:51:00 2023 -0500

    make generate output

commit a46b3ff9e5992fa96a337237f3fe704483e53a49
Author: gabemontero <gmontero@redhat.com>
Date:   Thu Mar 2 13:50:40 2023 -0500

    add proxy plugin crd
gmontero ~/go/src/github.com/codeready-toolchain/api  (add-proxy-config)$ make generate 
rm -rf config/
Re-generating the deepcopy go file & the Toolchain CRD files... 
/home/gmontero/go/src/github.com/codeready-toolchain/api/bin/controller-gen crd \
object paths="./..." output:crd:artifacts:config=config/crd/bases
re-generating the openapi go file...
/home/gmontero/go/src/github.com/codeready-toolchain/api/bin/openapi-gen --input-dirs ./api/v1alpha1/ \
--output-package github.com/codeready-toolchain/api/api/v1alpha1 \
--output-file-base zz_generated.openapi \
--go-header-file=make/go-header.txt
API rule violation: names_match,github.com/codeready-toolchain/api/api/v1alpha1,RegistrationServiceVerificationConfig,AWSSenderID
Dispatching CRD files in the 'host-operator' and 'member-operator' repositories...
# Now let's remove the CRDs from config/crd/bases directory
ls: cannot access 'config/crd/bases/*yaml': No such file or directory
Dispatch successfuly finished \o/
cp ../registration-service/deploy/registration-service.yaml ../host-operator/deploy/registration-service/registration-service.yaml
gmontero ~/go/src/github.com/codeready-toolchain/api  (add-proxy-config)$ git status 
On branch add-proxy-config
nothing to commit, working tree clean
gmontero ~/go/src/github.com/codeready-toolchain/api  (add-proxy-config)$
```

**Comment by gabemontero** [comment]
bump @alexeykazakov - can we rebase then merge this one ?

My thought was I would then look to make the sandbox-sre PRs as well as vendor this into my registration-service prototype and start refining that.

thanks

**Pull request merged**